File: num17.c
/*
If the numbers 1 to 5 are written out in words: one, two, three, four, five, then there are 3 + 3 + 5 + 4 + 4 = 19 letters used in total.

If all the numbers from 1 to 1000 (one thousand) inclusive were written out in words, how many letters would be used?


NOTE: Do not count spaces or hyphens. For example, 342 (three hundred and forty-two) contains 23 letters and 115 (one hundred and fifteen) contains 20 letters. The use of "and" when writing out numbers is in compliance with British usage.
*/

#include <stdio.h>

int main()
{
    int count[1001] = {0};
    count[1] = 3; // one
    count[2] = 3; // two
    count[3] = 5; // three
    count[4] = 4; // four
    count[5] = 4; // five
    count[6] = 3; // six
    count[7] = 5; // seven
    count[8] = 5; // eight
    count[9] = 4; // nine
    count[10] = 3; // ten
    count[11] = 6; // eleven
    count[12] = 6; // twelve
    count[13] = 8; // thirteen
    count[14] = 8; // fourteen
    count[15] = 7; // fifteen
    count[16] = 7; // sixteen
    count[17] = 9; // seventeen
    count[18] = 8; // eighteen
    count[19] = 8; // nineteen
    count[20] = 6; // twenty
    count[30] = 6; // thirty
    count[40] = 5; // forty
    count[50] = 5; // fifty
    count[60] = 5; // sixty
    count[70] = 7; // seventy
    count[80] = 6; // eighty
    count[90] = 6; // ninety
    count[100] = 10; // one hundred
    count[200] = 10; // two hundred
    count[300] = 12; // three hundred
    count[400] = 11; // four hundred
    count[500] = 11; // five hundred
    count[600] = 10; // six hundred
    count[700] = 12; // seven hundred
    count[800] = 12; // eight hundred
    count[900] = 11; // nine hundred
    count[1000] = 11; // one thousand
    
    int sum = 0;
        
    for(int i = 1; i <= 1000; i++)
    {
        if(count[i] == 0)                  // number is not in the upper count[]
        {
            if(i < 100)
            {
                count[i] = count[(i/10)*10] + count[i%10];
            }
            
            else
            {
                if(i % 100 == 0)
                {
                    count[i] = count[i];
                }
                else
                {
                    if(((i>110) && (i<120)) || ((i>210) && (i<220)) || ((i>310) && (i<320)) || ((i>410) && (i<420)) || ((i>510) && (i<520)) || ((i>610) && (i<620)) || ((i>710) && (i<720)) || ((i>810) && (i<820)) || ((i>910) && (i<920)))
                    {
                        count[i] = count[(i/100)*100] + count[(i%100)] + 3;                              // 3 means and
                    }
                    else
                    {
                        count[i] = count[(i/100)*100] + count[((i/10)%10)*10] + count[(i%100)%10] + 3;   // 3 means and
                    }
                }
            }
        }
    }
    
    for(int i = 1; i <= 1000; i++)
    {
        sum += count[i];
    }
    printf("%d\n", sum);
    return 0;
}

// answer : 21124

File: num17.py
dic={0:0,1:3,2:3,3:5,4:4,5:4,6:3,7:5,8:5,9:4,10:3,
     11:6,12:6,13:8,14:8,15:7,16:7,17:9,18:8,19:8,20:6,
     30:6,40:5,50:5,60:5,70:7,80:6,90:6,100:7,1000:11}
sum = 0

for i in range(1,1001):
    if len(str(i)) == 1:             # 1 digit
       sum += dic.get(i)

    elif len(str(i)) == 2:           # 10 digit
        if dic.get(i,0) == 0:        # dic.get(i,0) means if i can't be found in dic, get 0(ex:21,22,...etc)
            sum += dic.get(int(str(i)[0])*10)+dic.get(int(str(i)[1]))
        else :                        # (ex:10,11,12...etc)
            sum += dic.get(i)

    elif len(str(i)) == 3:           # 100 digit
        if i % 100 == 0:             # 100,200,... have no and
            num = dic.get(int(str(i)[0])) + dic.get(100)      # dic.get(100) is 7 (7 means hundred)
        else:                        # except 100,200,...
            num = dic.get(int(str(i)[0])) + dic.get(100) + 3  # plus and

        if dic.get(int(str(i)[1:]),0) == 0:                   # if 10 digit number is not in the dic
            sum += dic.get(int(str(i)[1])*10)+dic.get(int(str(i)[2])) + num
        else:                                                 # if 10 digit number is in the dic
            sum += dic.get(int(str(i)[1:])) + num

    else:                             # 1000
        sum += dic.get(i)

print(sum)